multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mmrcms/MMRCMS.kt

```diff
@@ -1,529 +1,487 @@
+                                Pair(it.toString(), it.toString())
+                            }.toTypedArray(),
                         ),
                     ),
                 )
+
+                if (tags.isNotEmpty()) {
+                    add(
+                        UriPartFilter(
+                            "Tag",
+                            "tag",
+                            arrayOf(
+                                "Any" to "",
+                                *tags.toTypedArray(),
+                            ),
+                        ),
+                    )
+                }
+
+                add(SortFilter())
             }
-            else -> FilterList(getInitialFilterList())
         }
+
+        return FilterList(filters)
     }
 
-    /**
-     * Class that creates a select filter. Each entry in the dropdown has a name and a display name.
-     * If an entry is selected it is appended as a query parameter onto the end of the URI.
-     * If `firstIsUnspecified` is set to true, if the first entry is selected, nothing will be appended on the the URI.
-     */
-    // vals: <name, display>
-    open class UriSelectFilter(
-        displayName: String,
-        private val uriParam: String,
-        private val vals: Array<Pair<String, String>>,
-        private val firstIsUnspecified: Boolean = true,
-        defaultValue: Int = 0,
-    ) :
-        Filter.Select<String>(displayName, vals.map { it.second }.toTypedArray(), defaultValue), UriFilter {
-        override fun addToUri(uri: Uri.Builder) {
-            if (state != 0 || !firstIsUnspecified) {
-                uri.appendQueryParameter(uriParam, vals[state].first)
+    private var categories = emptyList<Pair<String, String>>()
+
+    private var statuses = emptyList<Pair<String, String>>()
+
+    private var types = emptyList<Pair<String, String>>()
+
+    private var tags = emptyList<Pair<String, String>>()
+
+    private var fetchFiltersFailed = false
+
+    private var fetchFiltersAttempts = 0
+
+    private val fetchFiltersLock = ReentrantLock()
+
+    protected open fun fetchFilterOptions() {
+        Single.fromCallable {
+            if (!fetchFilterOptions) {
+                return@fromCallable
             }
-        }
-    }
 
-    class AuthorFilter : Filter.Text("Author"), UriFilter {
-        override fun addToUri(uri: Uri.Builder) {
-            uri.appendQueryParameter("author", state)
-        }
-    }
+            fetchFiltersLock.lock()
 
-    class SortFilter :
-        Filter.Sort(
-            "Sort",
-            sortables.map { it.second }.toTypedArray(),
-            Selection(0, true),
-        ),
-        UriFilter {
-        override fun addToUri(uri: Uri.Builder) {
-            uri.appendQueryParameter("sortBy", sortables[state!!.index].first)
-            uri.appendQueryParameter("asc", state!!.ascending.toString())
-        }
+            if (fetchFiltersAttempts > 3 || (fetchFiltersAttempts > 0 && !fetchFiltersFailed)) {
```

Comment:
Should check this condition before locking too, I guess.

Reply:
That was originally how I wrote it, but I realized these may be mutated by a previous call to `fetchFilterOptions()` if the user is quick enough.